query/common.d.ts

```diff
@@ -0,0 +1,287 @@
+export interface AvailableOrdersMetadataOpts { availableOrders: true; }
+
+export type ConditionalMetadataType<AdditionalMetadataType, M, OrderItemsType extends QuadTermName | RDF.Variable> = AdditionalMetadataType
+  & (M extends CardinalityMetadataOpts ? { cardinality: QueryResultCardinality } : Record<string, unknown>)
+  & (M extends OrderMetadataOpts ? { order: QueryOperationOrder<OrderItemsType>['terms'] } : Record<string, unknown>)
+  & (M extends AvailableOrdersMetadataOpts ? { availableOrders: QueryOperationOrder<OrderItemsType>[] } : Record<string, unknown>);
+
+/**
+ * Options that can be passed when executing a query.
+ */
+export interface QueryExecuteOptions<OrderItemsType extends QuadTermName | RDF.Variable> {
+  /**
+   * The required order for the result stream.
+   */
+  order?: QueryOperationOrder<OrderItemsType>;
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * Generic interface that defines query objects following the Future pattern.
+ */
+export interface BaseQuery {
+  /**
+   * Identifier for the type of result of tis query.
+   */
+  resultType: string;
+
+  /**
+   * Returns either a stream containing all the items that match the given query,
+   * a boolean or void depending on the semantics of the given query.
+   */
+  execute(opts?: any): Promise<ResultStream<any> | boolean | void>;
+}
+
+/**
+ * Query object that returns bindings.
+ */
+export interface QueryBindings<SupportedMetadataType> extends BaseQuery, BaseMetadataQuery<RDF.Variable, { variables: RDF.Variable[] }, SupportedMetadataType> {
+  resultType: 'bindings';
+  execute(opts?: QueryExecuteOptions<RDF.Variable>): Promise<ResultStream<Bindings>>;
+}
+
+/**
+ * Query object that returns quads.
+ */
+export interface QueryQuads<SupportedMetadataType> extends BaseQuery, BaseMetadataQuery<QuadTermName, unknown, SupportedMetadataType> {
+  resultType: 'quads';
+  execute(opts?: QueryExecuteOptions<QuadTermName>): Promise<ResultStream<RDF.Quad>>;
+}
+
+/**
+ * Query object that returns a boolean.
+ */
+export interface QueryBoolean extends BaseQuery {
+  resultType: 'boolean';
+  execute(): Promise<boolean>;
+}
+
+/**
+ * Query object that returns void.
+ */
+export interface QueryVoid extends BaseQuery {
+  resultType: 'void';
+  execute(): Promise<void>;
+}
+
+/**
+ * Union type for the different query types.
+ */
+export type Query<SupportedMetadataType> = QueryBindings<SupportedMetadataType> | QueryBoolean | QueryQuads<SupportedMetadataType> | QueryVoid;
+
+/**
+ * Bindings represents the mapping of variables to RDF values using an immutable Map-like representation.
+ * This means that methods such as `set` and `delete` do not modify this instance,
+ * but they return a new Bindings instance that contains the modification.
+ *
+ * Bindings instances are created using a BindingsFactory.
+ *
+ * The internal order of variable-value entries is undefined.
+ */
+export interface Bindings extends Iterable<[RDF.Variable, RDF.Term]> {
```

Comment:
>> it merely gives an illusion of type safety rather than actual type safety

> This pretty much sums up all of TypeScript, doesn't it? :)

Well, yes and no. TypeScript disappears at runtime and I agree it's not enough on its own to *guarantee* type safety if a program does any kind of input/output. However, one can use different strategies to ensure some level of safety. Reading from a relational database, for example, provides for some level of safety as long as the types are kept in sync with the schema, either manually or with some automation. If no schema is guaranteed at the source level we can always use validation.

However, a JS spec working on graphs where the only thing that can be taken for granted is the `S P O (G)` structure offers very little in terms of type safety and I fear that providing an easy/implicit way to cast query results into specific shapes might encourage a misplaced sense of security.

All that said, I do understand the argument for convenience. What if we were to offer a dedicated `TypedBindings` interfaces that extends the untype one, leaving the responsibility of casting one into the other to users?

```typescript
import { Bindings } from './query';
import { Variable, Term } from './data-model';

// This would have to become the new `Variable`. Setting `Value` to `string`
// by default should make this a non-breaking change.
interface TypedVariable<Value extends string = string> extends Variable {
  value: Value;
}

interface TypedBindings<Shape extends { [key: string]: Term }> extends Bindings {
  get<Key extends Extract<keyof Shape, string>>(key: Key): Shape[Key];
  get<Key extends Extract<keyof Shape, string>>(variable: TypedVariable<Key>): Shape[Key];
}
```

With this, one could easily do `(someBindingsObj as TypedBindings<SomeShape>)` . As per your comment,  it would require making `Variable` generic.

Reply:
It seems to me like quite some (discussion) work will have to go into defining such typed bindings. In any case, this would result in a breaking change for the current typings, so we probably don't want to do this in the near future. Given this, shall we move this discussion to a new issue?

In any case, I will look into adding a string-based `get` method on `Bindings`, as I agree that it might be more convenient in some cases.

Reply:
I agree with filing this into a separate issue. Typed bindings may take a lot of discussion indeed.